preprocessing.py

```diff
@@ -119,25 +119,24 @@ def read_tf_records(batch_size, tf_records, num_repeats=1,
     record_list = tf.data.Dataset.from_tensor_slices(tf_records)
 
     # compression_type here must agree with write_tf_examples
-    # cycle_length = how many tfrecord files are read in parallel
-    # block_length = how many tf.Examples are read from each file before
-    #   moving to the next file
-    # The idea is to shuffle both the order of the files being read,
-    # and the examples being read from the files.
-    if sloppy_interleave:
+    map_func = functools.partial(
+        tf.data.TFRecordDataset,
+        buffer_size=8 * 1024 * 1024,
+        compression_type='ZLIB')
+
+    if interleave:
+        # cycle_length = how many tfrecord files are read in parallel
+        # The idea is to shuffle both the order of the files being read,
+        # and the examples being read from the files.
         dataset = record_list.apply(tf.contrib.data.parallel_interleave(
-            functools.partial(tf.data.TFRecordDataset,
-                              buffer_size=8 * 1024 * 1024,
-                              compression_type='ZLIB'),
-            cycle_length=64, sloppy=True))
+            map_func, cycle_length=64, sloppy=True))
     else:
-        dataset = record_list.interleave(lambda x:
-                                         tf.data.TFRecordDataset(
-                                             x, compression_type='ZLIB'),
-                                         cycle_length=64, block_length=16)
+        dataset = record_list.flat_map(map_func)
+
     if filter_amount < 1.0:
-        dataset = dataset.filter(lambda x: tf.less(
-            tf.random_uniform([1]), filter_amount)[0])
+        dataset = dataset.filter(
+            lambda _: tf.random_uniform([]) < filter_amount)
```

Comment:
does this actually do shape inference to find the right size for tf.random_uniform? That seems pretty cool. 

Curious if this also fixes #520

Reply:
I think it does something different than you expect.
[] is a scalar random and filter get's run on every item in the dataset.

---
```
x = np.random.sample((10, 2)) 
dataset = tf.data.Dataset.from_tensor_slices(x)
iter = dataset.make_one_shot_iterator().get_next()
with tf.Session() as sess:
  while True:
    print(sess.run(iter))
iter = dataset.filter(lambda x: tf.random_uniform([]) < .3).make_one_shot_iterator().get_next()
with tf.Session() as sess:
  while True:
    print(sess.run(iter))
```

Reply:
oh, okay - never mind then :)